### DISCUSSION TITLE
Question about how to create a new note from a PDF

### DISCUSSION THREAD

**Discussion by kazdonkai** [RyotaUshio/obsidian-pdf-plus#86] [Q&A]
PDFから新しいノートを作成する方法についての質問です。
最終的には，これを一番少ない手順で実現したいです。
なので，上段のメニューに，”新しいノートの作成”というメニューがあって，すでにノートがない場合には，設定のテンプレートにしたがって，新しいノートを作成してほしいです。

ただし，では現在どのように作るのかについてですが，以下それについての質問です。

質問(1) 上段のプルダウンメニューにCreate new note がありますが，これはどのように使うのか？
①このメニューで，Create new noteにしておく。
②PDFの表題を範囲設定して，上のマーカーをクリックすると，新しいノートができました。
(画像の例では，｢日明関係史研究の最前線と教科書｣まで選択したので，ファイル名もその通りになりました)。

疑問：これと，設定との整合性はどのようになっているのか？
(a)New note title format にfile としたが，file名(｢日明関係史研究の最前線と教科書記述｣)ではノートが作成されない。
(b)Template で，テンプレートファイルを作って，例にしたがってproperty を指定しておいたが，反映されない。

とりあえず，(1)のみの質問です。
<img width="1642" alt="スクリーンショット 2024-03-05 13 52 53" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/90127278/611b45f6-0c04-43c0-99c4-e1c18d8b98e7">

**Comment by RyotaUshio** [comment]
ありがとうございます。回答の前に前提知識を確認したいのですが、"Quote", "Link Only", ..., "Create new note" などの項目を含んでいるドロップダウンメニューの内容は以下の設定で決まっているというのはご存知ですか？

<img width="781" alt="image" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/ddd3cc49-b8b1-46e6-9ca4-075f35bd23ba">

**Reply by RyotaUshio** [reply]
この設定では、リンクをコピーするときのフォーマットを、テンプレート機能を使って自由に指定することができます。
このテンプレート機能は自由度がかなり高く、`{{...}}`内では任意のJavaScriptを実行できるのですが、この機能を使うと
こんなことさえもできるんだよ、というひとつの例として、少し前のバージョンまでは "Create new note" という項目を入れていました。

これは、お試しいただいた通り、選択範囲をタイトルとし、選択範囲へのリンクを内容としてもつマークダウンファイルを新規作成するという処理をテンプレート内のJavaScriptによって実行するものです。（もはやフォーマットという枠を超えていますが、あくまでこんなこともできるよという例の提示です）

「どのように使うのか？」というご質問でしたが、あくまで例として入れているだけなので、（"Create new note"以外の項目も含めて）デフォルトのフォーマットについてはそのまま使ってもらうことはそもそもあまり想定していません。
フォーマットは自由に編集・削除・新規作成できるので、デフォルトの設定を参考にして、各ユーザに必要に応じて自分に合うフォーマットを設定し直してほしいと思っています。（ただ、デフォルトのままでも特に不自由なく使えるとは思います）

しかし今はあまりに高度な例を入れるとかえって混乱を生むだろうと重い、"Create new note"はデフォルトから除外しています。
（Kazdonkaiさんは除外前から使っていただいているので、古いデフォルト設定がそのまま残ってしまっているということです。）

---

"New note title format"などの設定は、設定画面をよく見ていただくとわかると思いますが、"PDF++: Create new note for auto-focus or auto-paste"というコマンドのための設定です。ドロップダウン内の"Create new note"とは関係はありません。

このコマンドを使って新規ノート作成すれば、以下の設定の内容が反映されるはずです。

- このコマンドを"Command to execute"オプションで指定すると、auto-focusまたはauto-paste時にペースト先のマークダウンファイルが特定できない場合（はじめてペーストする場合など）にこのコマンドが自動的に実行され、新規ノートがテンプレートにしたがって作成されます。
- Auto-focus/auto-pasteとは無関係に、現在開いているPDFファイルに対応するマークダウンファイルを新規作成するために手動で実行することもできます。その場合は、Commanderプラグインを使って、このコマンドをリボンアイコンやタブバーに割り当てるのが個人的なおすすめです。

<img width="801" alt="image" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/df02e4dd-5655-49ea-ab95-d0a7cf72a8a3">

**Comment by kazdonkai** [comment]
ありがとうございます。

疑問①について。
ちゃんと理解できたかどうか心許ないのですが，「私が試してみたドロップダウンメニューにあった"Creat new note "の設定では、いつかのバージョンで消してしまったが、わたしの方では、Jsonファイルに残ったままになっていたのでうまく機能しなかった｣ということでしょうか？
そうかなと思って、前にXで質問したときと同様に，Jsonファイルを消去して再インストールをしたら、メニューから消えました。

このドロップダウンによる新規作成については，それよりもコマンドを使った方がよいとのことなので、そちらの方法にしますが、
①コマンドによる作成方法と、すでに設定からはなくなったけれどもドロップダウンによる新規作成が｢関係はない｣ということが，名称からも似ていてユーザーにとってはわからないし混乱を招いた。
②いつかの時点で，その設定がなくなったけれども，前から使っているユーザーにとっては、Jsonファイルがそのままのために作者がやめたにも関わらず残ってしまう。
という問題があったわけです。
この問題については，バージョンを上げたことによる仕様の変更が，なんらかのプラグインにきちんと反映されて欲しい。
あるいはそれが無理であれば、リリースノートで、この仕様はなくした、あるいは変更したので、Jsonファイルの消去を促すなどの告知をしていただきたいと思います。

上を踏まえて，コマンドで"Pdf++:Create new note for auto-focus or auto-pase"
 をもちいて，新規ファイルを作成しましたが，画像のようになります。
(a)New note title formatの設定で、"file"としたことろ，"file"というノートが作成され，PDFのタイトル名にならない。
(b)Template ファイルを設定の"Example"をもとに改変して作成したところ，propertiesの部分はうまく機能しているが，本文がExampleのtemplateのままになっている。これは、最初に質問したときには，うまく機能していた(上の画像)。
今回は機能していないので質問します。

よろしくお願いします。
<img width="1297" alt="スクリーンショット 2024-03-09 21 38 24" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/90127278/eeb0d650-13fb-420c-857a-a8ed758fb87c">

**Reply by RyotaUshio** [reply]
> 「私が試してみたドロップダウンメニューにあった"Creat new note "の設定では、いつかのバージョンで消してしまったが、わたしの方では、Jsonファイルに残ったままになっていたのでうまく機能しなかった｣ということでしょうか？

はい、その理解であっています。（需要があるかわかりませんが、詳しく説明します：
PDF++を含めた多くのプラグインでは、「デフォルトの設定」というものが プラグインのコード(`main.js`)内に直接定義されています:

https://github.com/RyotaUshio/obsidian-pdf-plus/blob/754e949ea58db4afe037da415200e8ed6bc97f5e/src/settings.ts#L225

ユーザがはじめてプラグインをオンにすると、`data.json` にまずこのデフォルトの設定が書き込まれます。その後で、ユーザが設定画面から各種設定を変更すると、その度にその変更が `data.json` に書き込まれていきます。

プラグインがアップデートされると、この「デフォルトの設定」にも変更がある場合がありますが、ユーザがせっかく設定し直したものを新しいデフォルト設定によって上書きしては不都合が起きますので、既存のユーザの設定には特に触らないようになっています。

結果として、アップデート前からのユーザにはデフォルト設定の変更が反映されないようになっています（PDF++に限らず、ほぼすべてのプラグインでそうです）。
しかし、デフォルトの設定というのはあくまでおすすめ設定の一例の提示ですので、これが大きな問題であるとは考えていません。
）

> この問題については，バージョンを上げたことによる仕様の変更が，なんらかのプラグインにきちんと反映されて欲しい。
あるいはそれが無理であれば、リリースノートで、この仕様はなくした、あるいは変更したので、Jsonファイルの消去を促すなどの告知をしていただきたいと思います。

"Create new note"という名前がわかりにくかったというのは確かにそうだと思いますので今後気をつけます。
ただ、デフォルト設定というのはあくまでデフォルトの設定であり、おすすめの設定の例の提示でしかないので、デフォルト設定の変更が仕様の変更であるとは考えていません。また、先にご説明した通り、ユーザがせっかく設定し直したものを上書きすることは避けなければなりませんので、反映というのも難しいです。
デフォルトの設定＝おすすめ設定の一例が変わっただけですので、"Jsonファイルの消去を促す"ということは今後もしないかと思います（"Create new note"を消したのも、それに問題があるからではなく、単に最初に提示する例としては込み入りすぎているかなと思ったからでしかありません）。ご理解いただければと思います。



> (a)New note title formatの設定で、"file"としたことろ，"file"というノートが作成され，PDFのタイトル名にならない。

テンプレートの記法については設定画面に詳細な説明がありますので、そちらを見ていただければと思います。

<img width="777" alt="image" src="https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/c9450457-14c9-4217-8f22-96bf1580f6c8">

まず、`file`のような変数を使えるのは`{{  }}`内だけです。また、上の画像に説明されているように、`{{file}}`だけではファイル名にはならず
`{{file.basename}}`（拡張子なしファイル名）とか`{{file.name}}`（拡張子ありファイル名）とかとする必要があります。

例:
```
Comments for {{file.basename}}
```

> (b)Template ファイルを設定の"Example"をもとに改変して作成したところ，propertiesの部分はうまく機能しているが，本文がExampleのtemplateのままになっている。これは、最初に質問したときには，うまく機能していた(上の画像)。
今回は機能していないので質問します。

PDF++の設定画面内の説明でも触れていますが、Templaterの"Trigger Templater on new file creation"オプションはオンになっていますか？なっていれば、下の動画のようになるのではと思いますが、どうでしょうか。もしならない場合は、別のプラグインが干渉している可能性もあります。もし邪魔者のいない（PDF++とTempalterだけがオンになっている）sandbox vaultでも同様のことが起こる場合は、バグレポートを出していただければ調べてみますのでよろしくお願いします。

https://github.com/RyotaUshio/obsidian-pdf-plus/assets/72342591/5ee9851e-5be8-4427-9b74-b062832af59c

**Comment by kazdonkai** [comment]
詳しくありがとうございました。
おかげでうまくいくようになりました。

それについて，下で経過を説明しますが，その前提として述べておきたいことがあるので書きます。
Ushio さんには，ここでも、Xの方でも詳しく説明していただいていますが，その説明が開発者目線で，ユーザー目線になっていないことです。

わたしは開発者ではないので，論文などの文章を書く立場から比喩的に述べると、文章を書くときには，将来読んでくれる読者を想定して書く必要があるということです。(これは前に，noteに書いたことがあります)。

これを，プラグインに置き換えると，将来使ってくれるユーザーの目線から，システム設計や設定の説明を構成してほしいということです。

極端に言うと、できれば，Javascriptを知らなくても，あるいは，設定の説明を読まなくても，シームレスに使える仕様になってほしい，ということです。(これは極端な例で書いただけで、設定の説明を読めば最低限使うようになれるというのが現実的でしょうが・・)。

開発者目線で、ユーザー目線になっていないという例を挙げます。
たとえば、｢新しいノートを作成する｣という機能をめぐって、前には，プルダウンメニューとコマンドの2種類があったわけですが、｢プルダウンメニューは、1つの例として付けておいただけで、コマンドのノートの作成とは関係ない｣と説明なさいました。
たしかに、開発者目線としては、そのような説明になるのだろうし、プルダウンメニューから削除したのは、単なる例だから｢仕様｣の変更という訳ではないということでしょう。
しかし、ユーザーの立場から言えば、｢ある１つのプラグイン内で新しいノートを作る｣という機能があったとしたら、それが、プルダウンメニューからであろうと、コマンドであろうとおなじ働きをするのだろうと考えて実行するし、バージョンアップで、急にプルダウンメニューに残っているのに動作しなくてJSONを消して再インストールする必要があるのは仕様の変更ではないかと考えるわけです。


また文章に置き換えると，ある専門領域の研究者が読む専門の論文であれば、たとえば｢周知の様にメルロ＝ポンティは云々｣という説明をしたとしても当然理解されるわけですが、そうではなくて，｢メルロ＝ポンティ｣を知らない人に対してはそういう文章は書かないわけです。
つまり、前提的知識がない一般大衆が普及書を読む様に、極端に言えば，マニュアルを読まなくても電気製品を使うような感覚で，何もプログラム的な知識がなくても、プラグインを使える様な設計になってほしいということです。
また，このすぐれたプラグインが広がれば広がるほど、そういう人たちが増えていくでしょうし、そうであってこそユーザーが増えることになるでしょう。

このことを前提に、次の様に解決したということを書きます。

> (a)New note title formatの設定で、"file"としたことろ，"file"というノートが作成され，PDFのタイトル名にならない。

これは，設定画面のNew note title format の説明の
You can use the following variable: file 
 からそのままfile を入力したものです。
ここでは、file は変数で、Javascriptの記法にしたがって設定をしなければならない、ということが分かっていない無知の状態だった。

もちろん、このvariable については，設定画面で、
See below for the details about the variable 
となっていますが，実はこの変数の説明は下にはなくて，上の方の Link copy templates の下に
この変数はJavaScript の形式にしたがって云々と説明があって，
これを見ると，説明していただいたことがなるほどと理解できたわけです。そこで、
{{file.basename}}
と入力してうまく行ったわけですが、ここは設定部分の不手際(｢下にない｣)あるいは説明不足ではないかと思います。

このように，ユーザーが設定に書かれているすべての解説を丹念に見た上で，それを踏まえて入力すれば，期待通りの結果になりますが，わたしのような何も分かっていないユーザーもいるので，そこまで求めるのは酷というものではないでしょうか。

これをふまえて、ユーザー目線からの要望としては、たとえば初期設定として、PDFのファイル名が、自動的にノート名になる様に初期値として、{{file.basename}}が明示的に入力されていて、何もしないとそれが実行され、その上で、｢初期値を変更できます。それは、次の様なJavascriptの記法にしたがって下さい。｣云々というような説明にしてほしいわけです。


> (b)Template ファイルを設定の"Example"をもとに改変して作成したところ，propertiesの部分はうまく機能しているが，本文がExampleのtemplateのままになっている。

このTemplater を用いた本文のExample についての質問も上と同様で，
わたしは，プロパティの設定を使いたかったので，そのままExampleをコピーしてテンプレートファイルを作成したのですが、
Type note title の画面が出たわけですが，私としてはファイル名を更新する必要は無かったので，エスケープするとtemplate error になって，下の部分がそのまま残ってしまうと言う事態になっていたわけです。
<%* const title = await tp.system.prompt("Type note tile") -%>
<%* await tp.file.rename(title) %>
そこで，この部分をきちんと見てみたところ，新規ノートを作成したときに，ノートタイトルをあらためて入力する場合の例であることが分かりました。そこで、この記述の部分は省いて，本文は下のようにして対応しました。
# {{file.basename}}

ファイル名をノート名にしたいか、あるいは、このTemplateのように、新しいノート名を入力したいかは、ユーザーによって異なると思います。
ネット上では、省略したPDF名が多いですが、わたしはダウンロードするときに、極力、筆者・論文名を加えて、Obsidianでもそのままノート名に使える様なファイル名で保存しています。
ですから、そのようなユーザーにとっては、設定にサンプルとしてあげられているこのテンプレートは不要であることになります。
一方、短縮名をObsidianのノート名にするときに変更したい人にとっては有用なテンプレートです。
このようにユーザーの事情はさまざまなので、このテンプレートは変更用のテンプレートであるという説明を加えるとか、その上で、選択肢の1つにするとかにしてほしいと思います。

ということで、このスレッドの質問は解決しましたが、以上で述べたかったことの意を汲んでいただければ幸いです。
(また、ユーザー目線からの解説にしていただければ、開発者の意図を説明するという立場からの解説もそんなにしなくてもよいことにもなるのではないでしょうか😀)
